### PR TITLE
Bug 1157737 - Use MutationObserver to detect that a form has been injected

### PR DESCRIPTION
**Pull request by fluffyemily**
https://bugzilla.mozilla.org/show_bug.cgi?id=1157737

Question: Is searching the DOM for a form injected be and then calling `findLogins` more efficient than just using `findLogins` and possibly executing it _every_ time there is a mutation regardless of whether a form has been injected?

### CODE REVIEW

File: Client/Assets/LoginsHelper.js [hunk 1]
```diff
@@ -599,19 +599,28 @@ function onBlur(event) {
   LoginManagerContent.onUsernameInput(event)
 }
 
-window.addEventListener("load", function(event) {
+var documentBody = document.querySelector('body')
```
Comment (thebnich) [inline Client/Assets/LoginsHelper.js]: Some drive-by comments: you can just use `document.body` here.

File: Client/Assets/LoginsHelper.js [hunk 2]
```diff
@@ -599,19 +599,28 @@ function onBlur(event) {
...
+var observer = new MutationObserver(function(mutations) {
+    findLogins()
+});
+
+observer.observe(documentBody, { attributes: false, childList: true, characterData: false });
+
+function findLogins(event) {
   try {
     for (var i = 0; i < document.forms.length; i++) {
       LoginManagerContent._asyncFindLogins(document.forms[i], { })
```
Comment (thebnich) [inline Client/Assets/LoginsHelper.js]: By looping over all forms on the page, it looks like you'll end up re-triggering the query for all forms every time a new one is added. Instead, you can use the `MutationObserver` to filter out `form`-specific additions, and ignore other events. For example:

```
var observer = new MutationObserver(function(mutations) {
  for (var i = 0; i < mutations.addedNodes.length; i++) {
    var node = mutations.addedNodes[i];
    if (node.tagName === "FORM") {
      LoginManagerContent._asyncFindLogins(node, { });
    }
  }
});
```

File: Client/Assets/LoginsHelper.js [hunk 3]
```diff
@@ -599,19 +599,28 @@ function onBlur(event) {
   LoginManagerContent.onUsernameInput(event)
 }
 
-window.addEventListener("load", function(event) {
+var documentBody = document.querySelector('body')
+var observer = new MutationObserver(function(mutations) {
+    findLogins()
+});
+
+observer.observe(documentBody, { attributes: false, childList: true, characterData: false });
```
Comment (thebnich) [inline Client/Assets/LoginsHelper.js]: I wonder if you also want to include `subtree: true` here? From the docs, it looks like `childList` might just be direct descendants. Worth testing, though (speaking of which, it might be a good idea to add a simple test case for dynamic `form` insertions).

File: Client/Assets/LoginsHelper.js [hunk 4]
```diff
@@ -599,19 +599,28 @@ function onBlur(event) {
   LoginManagerContent.onUsernameInput(event)
 }
 
-window.addEventListener("load", function(event) {
+var documentBody = document.querySelector('body')
+var observer = new MutationObserver(function(mutations) {
+    findLogins()
```
Comment (thebnich) [inline Client/Assets/LoginsHelper.js]: Nits: 2-space indentation in JS files, and semicolon at end of line

File: Client/Assets/LoginsHelper.js (hunk 2 above)
Comment (jhugman) [inline Client/Assets/LoginsHelper.js]: I like this, though I worry that we'd not catch forms that were assembled in a peculiar/pathological(?) order – e.g. add a FORM, add a INPUT & PASSWORD to that form.

**Comment by jhugman** [comment]
This could be more efficient thebnich approach, but it's correct.

File: Client/Assets/LoginsHelper.js (hunk 2 above)
Comment (thebnich) [inline Client/Assets/LoginsHelper.js]: Yeah, I had considered the same...my initial thought was that we should be fine since `MutationObserver` is async, but perhaps we could throw in a `setTimeout` to wait for the next tick, if needed.

Comment (fluffyemily) [inline Client/Assets/LoginsHelper.js]: So, I started out with @thebnich's solution, then moved to this one, then couldn't decide which was best and so added the question in the top. On, say, the instagram login page, the mutation observer gets called 6 times, one one of those times does a form get injected. This means that we are doing 6 DOM searches for the form tag (I found that simply searching the added nodes doesn't consistently find the FORM element) over 2 executions of the findLogins. But this might be different in other webpages. 

I'll move to searching for the FORM element though.

File: Client/Assets/LoginsHelper.js (hunk 3 above)
Comment (fluffyemily) [inline Client/Assets/LoginsHelper.js]: I found that adding subtree: true made no difference to the addedNodes list provided. Unless the FORM is the root node, then no form is found whether subtree is present or not
